wgengine/monitor/monitor_windows.go

```diff
@@ -7,273 +7,159 @@ package monitor
 import (
 	"context"
 	"errors"
-	"fmt"
-	"runtime"
+	"golang.zx2c4.com/wireguard/windows/tunnel/winipcfg"
 	"sync"
-	"syscall"
-	"time"
-	"unsafe"
-
-	"golang.org/x/sys/windows"
-	"tailscale.com/net/interfaces"
 	"tailscale.com/types/logger"
-)
-
-const (
-	pollIntervalSlow = 15 * time.Second
-	pollIntervalFast = 3 * time.Second
-	pollFastFor      = 30 * time.Second
+	"time"
 )
 
 var (
-	iphlpapi                 = syscall.NewLazyDLL("iphlpapi.dll")
-	notifyAddrChangeProc     = iphlpapi.NewProc("NotifyAddrChange")
-	notifyRouteChangeProc    = iphlpapi.NewProc("NotifyRouteChange")
-	cancelIPChangeNotifyProc = iphlpapi.NewProc("CancelIPChangeNotify")
-)
-
-const (
-	_STATUS_PENDING = 0x00000103 // 259
-	_STATUS_WAIT_0  = 0
+	errClosed = errors.New("closed")
 )
 
 type unspecifiedMessage struct{}
 
 func (unspecifiedMessage) ignore() bool { return false }
 
-type pollStateChangedMessage struct{}
-
-func (pollStateChangedMessage) ignore() bool { return false }
-
 type messageOrError struct {
 	message
 	error
+	evt string
 }
 
 type winMon struct {
-	ctx             context.Context
-	cancel          context.CancelFunc
-	messagec        chan messageOrError
-	logf            logger.Logf
-	pollTicker      *time.Ticker
-	lastState       *interfaces.State
-	closeHandle     windows.Handle // signaled upon close
-	goroutineDoneCh chan struct{}  // closed when awaitIPAndRouteChanges goroutine has stopped
-
-	mu            sync.Mutex
-	lastNetChange time.Time
-	inFastPoll    bool // recent net change event made us go into fast polling mode (to detect proxy changes)
+	logf                  logger.Logf
+	ctx                   context.Context
+	mu                    sync.Mutex // guards following
+	addressChangeCallback *winipcfg.UnicastAddressChangeCallback
+	routeChangeCallback   *winipcfg.RouteChangeCallback
+	messagec              chan messageOrError
 }
 
 func newOSMon(logf logger.Logf, _ *Mon) (osMon, error) {
-	closeHandle, err := windows.CreateEvent(nil, 1 /* manual reset */, 0 /* unsignaled */, nil /* no name */)
-	if err != nil {
-		return nil, fmt.Errorf("CreateEvent: %w", err)
+	m := &winMon{
+		logf: logger.WithPrefix(logf, "monitor_windows: "),
```

Comment:
no need for the prefix; the caller already puts "monitor: " before them all, and windows is obvious from context. (it's global for a whole log file)

Reply:
removed